modules/internal/ChapelArray.chpl

```diff
@@ -309,33 +305,74 @@ module ChapelArray {
                       " If so, use {...} instead of [...].");
     }
 
+    param homog = isHomogeneousTuple(elems);
+
     // elements of string literals are assumed to be of type string
-    type eltType = _getLiteralType(elems(0).type);
+    type eltType = if homog then elems(0).type
+                            else chpl_computeUnifiedType(elems);
+
     var dom = {arrayLiteralLowBound..#k};
     var arr = dom.buildArray(eltType, initElts=false);
 
-    for param i in 0..k-1 {
-      type currType = _getLiteralType(elems(i).type);
-
-      if currType != eltType {
-        compilerError( "Array literal element " + i:string +
-                       " expected to be of type " + eltType:string +
-                       " but is of type " + currType:string );
+    if homog {
+      for i in 0..<k {
+        ref dst = arr(i+arrayLiteralLowBound);
+        ref src = elems(i);
+        __primitive("=", dst, src);
+      }
+    } else {
+      for param i in 0..k-1 {
+        type currType = elems(i).type;
+
+        ref dst = arr(i+arrayLiteralLowBound);
+        ref src = elems(i);
+        if (currType == eltType ||
+            Reflection.canResolve("=", dst, src)) {
+          __primitive("=", dst, src);
```

Comment:
I don't know what primitive assignment will or won't handle, and just used it because it was what was used here before my changes, and worked for the int/real case, as you note.  I'm open to adding more testing, but what sort of non-trivial elements would you like me to test it with?